lib/ansible/modules/cloud/amazon/lambda.py

```diff
@@ -189,30 +192,53 @@
       }
 '''
 
-# Import from Python standard library
+from ansible.module_utils._text import to_native
+from ansible.module_utils.aws.core import AnsibleAWSModule
+from ansible.module_utils.ec2 import get_aws_connection_info, boto3_conn, camel_dict_to_snake_dict
 import base64
 import hashlib
+import traceback
 
 try:
-    import botocore
-    HAS_BOTOCORE = True
+    from botocore.exceptions import ClientError, ValidationError, ParamValidationError
 except ImportError:
-    HAS_BOTOCORE = False
+    pass  # protected by AnsibleAWSModule
 
-try:
-    import boto3
-    HAS_BOTO3 = True
-except ImportError:
-    HAS_BOTO3 = False
+
+def get_account_id(module, region=None, endpoint=None, **aws_connect_kwargs):
+    """return the account id we are currently working on
+
+    get_account_id tries too find out the account that we are working
+    on.  It's not guaranteed that this will be easy so we try in
+    several different ways.  Giving either IAM or STS privilages to
+    the account should be enough to permit this.
+    """
+    try:
+        sts_client = boto3_conn(module, conn_type='client', resource='sts',
+                                region=region, endpoint=endpoint, **aws_connect_kwargs)
+        account_id = sts_client.get_caller_identity().get('Account')
+    except ClientError:
+        try:
+            iam_client = boto3_conn(module, conn_type='client', resource='iam',
+                                    region=region, endpoint=endpoint, **aws_connect_kwargs)
+            account_id = iam_client.get_user()['User']['Arn'].split(':')[4]
+        except ClientError as e:
+            if (e.response['Error']['Code'] == 'AccessDenied'):
+                except_msg = to_native(e.message)
+                account_id = except_msg.search("arn:aws:iam::([0-9]{12,32}):user/").group(1)
```

Comment:
think this _is_ currently working with temporary credits but your fix seems more general so I will go for it.